src/mono/wasm/runtime/logging.ts

```diff
@@ -0,0 +1,207 @@
+regexes.push(/(?<replaceSection>[a-z]+:\/\/[^ )]*:wasm-function\[(?<funcNum>\d+)\]:0x[a-fA-F\d]+)/);
+
+//# <?>.wasm-function[8962]
+regexes.push(/(?<replaceSection><[^ >]+>[.:]wasm-function\[(?<funcNum>[0-9]+)\])/);
+
+export function mono_wasm_symbolicate_string(message: string): string {
+    try {
+        if (wasm_func_map.size == 0)
+            return message;
+
+        const origMessage = message;
+
+        for (let i = 0; i < regexes.length; i++) {
+            const newRaw = message.replace(new RegExp(regexes[i], "g"), (substring, ...args) => {
+                const groups = args.find(arg => {
+                    return typeof (arg) == "object" && arg.replaceSection !== undefined;
+                });
+
+                if (groups === undefined)
+                    return substring;
+
+                const funcNum = groups.funcNum;
+                const replaceSection = groups.replaceSection;
+                const name = wasm_func_map.get(Number(funcNum));
+
+                if (name === undefined)
+                    return substring;
+
+                return substring.replace(replaceSection, `${name} (${replaceSection})`);
+            });
+
+            if (newRaw !== origMessage)
+                return newRaw;
+        }
+
+        return origMessage;
+    } catch (error) {
+        console.debug(`MONO_WASM: failed to symbolicate: ${error}`);
+        return message;
+    }
+}
+
+export function mono_wasm_stringify_as_error_with_stack(err: Error | string): string {
+    let errObj: any = err;
+    if (!(err instanceof Error))
+        errObj = new Error(err);
+
+    // Error
+    return mono_wasm_symbolicate_string(errObj.stack);
+}
+
+export function mono_wasm_trace_logger(log_domain_ptr: CharPtr, log_level_ptr: CharPtr, message_ptr: CharPtr, fatal: number, user_data: VoidPtr): void {
+    const origMessage = Module.UTF8ToString(message_ptr);
+    const isFatal = !!fatal;
+    const domain = Module.UTF8ToString(log_domain_ptr);
+    const dataPtr = user_data;
+    const log_level = Module.UTF8ToString(log_level_ptr);
+
+    const message = `[MONO] ${origMessage}`;
+
+    if (INTERNAL["logging"] && typeof INTERNAL.logging["trace"] === "function") {
+        INTERNAL.logging.trace(domain, log_level, message, isFatal, dataPtr);
+        return;
+    }
+
+    switch (log_level) {
+        case "critical":
+        case "error":
+            console.error(mono_wasm_stringify_as_error_with_stack(message));
+            break;
+        case "warning":
+            console.warn(message);
+            break;
+        case "message":
+            console.log(message);
+            break;
+        case "info":
+            console.info(message);
+            break;
+        case "debug":
+            console.debug(message);
+            break;
+        default:
+            console.log(message);
+            break;
+    }
+}
+
+export function setup_proxy_console(id: string, console: any, origin: string): void {
```

Comment:
Why is this `any`? Shouldn't it be:

```suggestion
export function setup_proxy_console(id: string, console: Console, origin: string): void {
```

Reply:
`console[m] = proxyConsoleMethod` this assignment would need to fully match the type.

Reply:
I cast it to any bit later, so that method signature could be better, but the ts compiler is ok to assign to that object.